### PR TITLE
add CODEOWNERS

### LINKED ISSUE #9340: rfc: use codeowners to automatically request reviews
Currently, we're frequently requesting reviews to maintainers, seemingly randomly, to get pull requests looked at sooner.

I'd like to propose that we reduce this practice, in favor of a more opt-in approach where we use a CODEOWNERS file to automatically request reviews on pages based on the maintainers interests.

The idea is that, if someone is primarily interested in reviewing pages for a certain language, operating system, or maybe the GitHub health files, they can opt-in for that.

This has a few benefits:
1. The active reviewers will no longer have to request reviews from other maintainers as frequently.
2. Less active reviewers (such as myself) may be more inclined to respond to review requests when it's something we're actively interested in.

Docs:
https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/customizing-your-repository/about-code-owners

The file would look something like this:

```txt
# Default
* {whoever is interested in anything not otherwise matched by anything below}

# Languages
/pages.de/ @tldr-pages/german 
/pages.es/ @tldr-pages/spanish 

# Language / Operating System
linux/ {whoever is interested in all linux pages regardless of language}

/pages/**/linux/ {whoever is interested in english pages for linux tools}
/pages/**/common/ {whoever is interested in english pages for common tools} 

# Pages
/pages/**/register_new_matrix_user.md @SethFalco 

# Health Files
/*.md @SethFalco 
/contributing-guides/*.md @SethFalco 

# Dependencies
/*.json {whoever is interested in linters/ci}
/requirements.txt {whoever is interested in linters/ci}
```
> Note that each line in the CODEOWNERS file can have more than one team/user in it.

We can always play with this whenever we want, or remove names when someone wants to opt out of being requested for review. We can't allow the file to exceed 3 MB, so we can't have too much fun with page-specific code owners, but overall, I don't expect that to be a problem.

The only caveat I see is that with how this is handled, if something further down matches a code owner, none of the matches above will count. When defining this file, I recommend starting with the most abstract file groups and working your way to be more specific.

Note above how I handle the `Language / Operating System` section in the example. I specify the OS first, then only after, language-specific versions for the OS. If it was written the other way around, then `/pages/**/linux/` would be redundant as it'd _always_ match `linux/` afterwards which would steal the review.

This also means, however, that the global reviewers at the top will not be requested to review if something further down matches. (Unless that global reviewer is also in the list of the last matched glob.)

What are our overall thoughts on this?

While it's not as clean as I'd like, I think this is a solid approach as it gives maintainers more control over when they're requested to review tasks.

None of this means that we can't request reviews from other maintainers anymore. Just that hopefully we can automate some of it away so maintainers get the things more interesting to them, and we can fallback to manual requests if things aren't getting reviewed.

### PR DESCRIPTION
**Pull request by pixelcmtd**
Closes #9340.

### CODE REVIEW

**Review by kbdharun — COMMENTED** [review]
~~I have a doubt does this auto request reviews from code owners or does it also set branch protection preventing other collaborators from merging the PRs.~~

Edit: I remember now that I asked this https://github.com/tldr-pages/tldr/issues/9340#issuecomment-1294644940 before so this is fine IG. Thanks for your contribution.

**Review by kbdharun — APPROVED** [review]

**Review by SethFalco — APPROVED** [review]

**Review by mebeim — APPROVED** [review]

**Comment by waldyrious** [comment]
Can we add `pages.pt_BR` and `pages.pt_PT` to a `portuguese` team? I assume that would be just me for now, but at least it sets up the scaffolding for a proper team later on.

**Comment by kbdharun** [comment]
Nice, Can you invite @waldyrious to it?

**Comment by pixelcmtd** [comment]
> Nice, Can you invite @ waldyrious to it?

I have already done that.

**Comment by acuteenvy** [comment]
Is it possible to add non-collaborators as codeowners? If so, could you add me to `pages.pl`?

**Comment by SethFalco** [comment]
@acuteenvy 

> Is it possible to add non-collaborators as codeowners? If so, could you add me to pages.pl?

Unfortunately, we're not able to do that.

> The people you choose as code owners must have write permissions for the repository.
> 
> — https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/customizing-your-repository/about-code-owners

**Comment by pixelcmtd** [comment]
> Is it possible to add non-collaborators as codeowners? If so, could you add me to `pages.pl`?

@acuteenvy Since you've already committed quite a bit to `tldr`, I've opened #10187 to add you as a collaborator. Adding non-collaborators as code owners isn't possible, as @SethFalco already noted.

@ everyone Is there **any** way to add non-members to teams? Because I really hate the idea of mixing users and teams in the code owners file. I feel like, when we adopt code owners, the teams become kind of redundant, as they're practically only used by the GitHub review requesting thing. So we might just as well remove them then and just have people adding themselves as code owners in a PR. What do y'all think?

**Comment by kbdharun** [comment]
> @ everyone Is there **any** way to add non-members to teams? Because I really hate the idea of mixing users and teams in the code owners file. I feel like, when we adopt code owners, the teams become kind of redundant, as they're practically only used by the GitHub review requesting thing. So we might just as well remove them then and just have people adding themselves as code owners in a PR. What do y'all think?

But doesn't the codeowners file have a size limit?

Maybe we can have the current method but for non-collaborators we can add their names manually.

**Comment by pixelcmtd** [comment]
> But doesn't the codeowners file have a size limit of 3MB and wouldn't it fill up soon if we add usernames?

Do you even know how much a megabyte is? The list of all 44.2k current stargazers (yes, GitHub is weird) is less than 0.5MB (`gh api /repos/tldr-pages/tldr/stargazers --jq 'map(.["login"]) | join(",@")' --paginate | sed 's/^/@/' | tr '\n' ',' | sed 's/,@/, @/g' | wc -c` → 471740). I lack the level of imagination needed to envision a situation in which we'd ever run into that limit.

> An alternative is dropping the tldr repository (write access) from teams and changing it to triage so that we can review PRs. The codeowners will get notified anyways for the respective locales so I think we do not need the team request review in our side.

What exactly would be the advantage of that? I'm not able to fully parse the second sentence, but I also need some sleep 😅

**Comment by kbdharun** [comment]
> > But doesn't the codeowners file have a size limit of 3MB and wouldn't it fill up soon if we add usernames?
> 
> Do you even know how much a megabyte is? 

Yes

> The list of all 44.2k current stargazers (yes, GitHub is weird) is less than 0.5MB (`gh api /repos/tldr-pages/tldr/stargazers --jq 'map(.["login"]) | join(",@")' --paginate | sed 's/^/@/' | tr '\n' ',' | sed 's/,@/, @/g' | wc -c` → 471740). I lack the level of imagination needed to envision a situation in which we'd ever run into that limit.
> 

Sorry that was a bad take, what I was trying to mean was as we add more members wouldn't the file be cluttered even through the maximum file size is 3MB.

> > An alternative is dropping the tldr repository (write access) from teams and changing it to triage so that we can review PRs. The codeowners will get notified anyways for the respective locales so I think we do not need the team request review in our side.
> 

> What exactly would be the advantage of that? I'm not able to fully parse the second sentence, but I also need some sleep 😅

It is simple, we already invite collaborators to the repository. So an additional write access in team is not necessary, in chatroom some days ago you mentioned that we can't request reviews if they don't have access to repo, that's what triage comes in (it allows people in team to review PRs while not having write access to repository), in this way we can add non-collaborators to it. Since the team would be specified in codeowners file we won't need the manual request review anyway, if we need it then you can update the repo access in team to triage.

**Comment by pixelcmtd** [comment]
> > Do you even know how much a megabyte is?
> 
> Yes

Great

> Sorry that was a bad take, what I was trying to mean was as we add more members wouldn't the file be cluttered even through the maximum file size is 3MB.

It could become that, yes. But let's take it realistically: We currently have 45 collaborators (including org members/owners).

44 people with access, 1 invitation

Even if everyone was to suddenly add themselves to many files (that's unlikely), it could be something like 100 lines with 3 people each. That's a bit cluttered, but very much manageable and I'd personally prefer it over having to manage teams.

> It is simple, we already invite collaborators to the repository. So an additional write access in team is not necessary, in chatroom some days ago you mentioned that we can't request reviews if they don't have access to repo, that's where triage comes in (it allows people in team to review PRs while not having write access to repository), in this way we can add non-collaborators to it.
> 
> Since the team would be specified in the codeowners file we won't need the manual request review anyway so I suggested removing the repo access, if we need it then you can update the repo access in team to triage.

Yes, we could demote every team to just triage access and it should still work. But afaik, there is no way to join the teams without being a member of the org. (you remember, how I became an org member, right?) Are you suggesting, there is a way of getting around this?

**Comment by kbdharun** [comment]
> > It is simple, we already invite collaborators to the repository. So an additional write access in team is not necessary, in chatroom some days ago you mentioned that we can't request reviews if they don't have access to repo, that's where triage comes in (it allows people in team to review PRs while not having write access to
> > Since the team would be specified in the codeowners file we won't need the manual request review anyway so I suggested removing the repo access, if we need it then you can update the repo access in team to triage.
> 
> Yes, we could demote every team to just triage access and it should still work. But afaik, there is no way to join the teams without being a member of the org. (you remember, how I became an org member, right?) Are you suggesting, there is a way of getting around this?

Yeah, there is no alternative without being member of the org. We could set the team's members to private, but I think it will conflict with existing public members like me. So considering alternatives, seems like your suggested method with username would be better feasible.

**Comment by kbdharun** [comment]
@pixelcmtd Done, updated the file as suggested. Feel free to review the changes (Check the commit message for more info).

**Comment by kbdharun** [comment]
@nicokosi, @patricedenis, @Nico385412 Generally, when we have French translation PRs we request reviews from you guys, are you interested to be added to the CODEOWNERS file here? (Note: You will be auto-requested for review when there is a PR related to French pages)

File: .github/CODEOWNERS
Comment (SethFalco) [inline .github/CODEOWNERS:17]: You can chain multiple people on the same path, i.e.

```suggestion
/contributing-guides/*.it.md @mebeim @yutyo
```

Sources:

> The pattern is followed by one or more GitHub usernames or team names using the standard @username or @org/team-name format.
> 
> ```
> /scripts/ @doctocat @octocat
> ```
> — https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/customizing-your-repository/about-code-owners

Comment (kbdharun) [reply .github/CODEOWNERS:17]: That's a great idea, I will update the file now.

**Comment by einverne** [comment]
> Edit 1: Also, @einverne since you are a native speaker and review PRs in `pages.zh` are you interested to be added as a CODEOWNER for it?

Sure, I would be interested in being added as a CODEOWNER for the ﻿pages.zh directory.

**Review by blueskyson — APPROVED** [review]
This is cool.

**Comment by patricedenis** [comment]
Hi everyone.
I confess I'm not that active in code reviewing these days. I will try to be back there more frequently. But if you want to add me, I won't fight against it. :smiley:

**Comment by kbdharun** [comment]
Is this PR GTG? 

> Maybe we could add the code owners for French pages at a later date instead of prolonging this PRs merge.

**Pull request merged**
